server/src/api/gitLab/gitLabIntegration.ts

```diff
@@ -0,0 +1,213 @@
+import { OAuth2 } from 'oauth';
+import { randomBytes } from 'crypto';
+
+import {
+  ImportedTask,
+  TaskFilters,
+  IntegrationProvider,
+  IntegrationEntry,
+  IntegrationConfig,
+  InvalidTokenError,
+  InvalidGrantError,
+} from '../integration';
+
+const authorizePath = '/oauth/authorize';
+const tokenPath = '/oauth/token';
+const redirectBaseURL = `${process.env.FRONTEND_BASE_URL}/oauth/redirect`;
+
+const oauth2 = (key: string, secret: string, host: string) =>
+  new OAuth2(key, secret, host, authorizePath, tokenPath);
+
+export const GitLabIntegration: IntegrationEntry = {
+  defaultConfig: { host: 'gitlab.com' },
+  configOptions: [
+    { field: 'projectId' },
+    { field: 'consumerkey' },
+    { field: 'privatekey', secret: true },
+  ],
+  getIntegrationProvider: (config, tokens) =>
+    new GitLabImporter(config, tokens),
+  getOAuthProvider: (config) => ({
+    instance: config.host,
+    authorizationURL: (roadmapId) =>
+      new Promise((resolve, reject) => {
+        const state = randomBytes(10).toString('hex');
+
+        const redirectUri = encodeURIComponent(
+          `${redirectBaseURL}?roadmapId=${roadmapId}&integrationName=gitLab`,
+        );
+        resolve({
+          url: `https://${config.host}${authorizePath}?client_id=${config.consumerkey}&response_type=code&state=${state}&scope=read_api&redirect_uri=${redirectUri}`,
+          token: '',
+          tokenSecret: '',
+        });
+      }),
+    swapOAuthToken: ({ verifierToken, token, tokenSecret, roadmapId }) =>
+      new Promise((resolve, reject) => {
+        const redirectUri = `${redirectBaseURL}?roadmapId=${roadmapId}&integrationName=gitLab`;
+
+        oauth2(
+          config.consumerkey,
+          config.privatekey,
+          `https://${config.host}`,
+        ).getOAuthAccessToken(
+          verifierToken,
+          {
+            grant_type: 'authorization_code',
+            redirect_uri: redirectUri,
+          },
+          (err, access_token, refresh_token, result) => {
+            if (err) {
+              if (err.statusCode === 401 && /token/i.test(err.data))
+                reject(new InvalidTokenError(err.data));
+              else if (err.statusCode === 400 && /grant/i.test(err.data))
+                reject(new InvalidGrantError(err.data));
+              else reject(err);
+            } else
+              resolve({
+                accessToken: access_token,
+                refreshToken: refresh_token,
+              });
+          },
```

Comment:
```suggestion
          (err, accessToken, refreshToken) => {
            if (err) reject(convertError(err));
            else resolve({ accessToken, refreshToken });
          },
```

Couple suggestions:
- the error check would be good to extract to a function (something like below)
- the parameters can be freely named, so could use the field names directly, as those also are camelCase which is the style convention
- the unused (trailing) parameter should be removed

```ts
const convertError = (err: { statusCode: number; data?: any }) => {
  if (err.statusCode === 401 && /token/i.test(err.data))
    return new InvalidTokenError(err.data);
  if (err.statusCode === 400 && /grant/i.test(err.data))
    return new InvalidGrantError(err.data);
  return err;
};
```